disk_objectstore/utils.py

```diff
@@ -1519,3 +1534,117 @@ def merge_sorted(iterator1: Iterable[Any], iterator2: Iterable[Any]) -> Iterator
     for item, _ in detect_where_sorted(iterator1, iterator2):
         # Whereever it is (only left, only right, on both) I return the object.
         yield item
+
+
+def should_compress(
+    source_stream: StreamSeekBytesType,
+    compress_mode: CompressMode,
+    source_compressed: bool,
+    source_length: int,
+    source_size: int,
+) -> bool:
+    """Return a boolean determining if the source_stream should be compressed or not.
+
+    :param source_stream: a seekable stream for which we should decide if we need to compress or not.
+    :param compress_mode: the compress mode that should be honoured by should_compress (i.e., what
+       the user asked for). E.g.:
+
+       - If `compress_mode` is `CompressMode.YES`, the answer is always True.
+       - If `compress_mode` is `CompressMode.NO`, the answer is always False.
+       - If `compress_mode` is ``CompressMode.KEEP`, the answer is equal to the input value of `source_compressed`
+         (i.e., the compression is unchanged)
+       - If `compress_mode` is `CompressMode.AUTO`, the stream is analyzed in a not-too-expensive way to determine
+         whether compression should be applied or not (i.e., if compressing will give a benefit or not).
+    :param source_compressed: whether the source stream was already compressed
+    :param source_length: the length (on disk, i.e. the compressed size if source_compressed is True) of
+        the source_stream - can be used to quickly decide if compression is already efficient on this object.
+    :param source_size: the uncompressed size of the source_stream - togetehr with source_length, can be used to
+        quickly decide if compression is already efficient on this object.
+
+    Note that in some cases this function will have to read parts of the stream to do some heuristics, so the stream
+    needs to be readable and seekable (because at the end the stream will be put back in the same position as when
+    we entered this method).
+    """
+    # It shoudl at least reduce by 10% to be worth compressing
+    compression_threshold = 0.9
```

Comment:
Would it make sense for compression to should be turned off if the stream (file) is shorter than a certain length? 

```
 echo '' | gzip -f | wc
```

gives me an output length of 21. 
But maybe a threshold length of 4KB should be used? This way we speed up the access of small files where overhead of decompression could take a significant impact on the performance.

Reply:
if you don't put the gzip headers, it's only 8 bytes:
```python
import zlib; print(len(zlib.compress(b'')))
```
returns 8.

Anyway, it's true that small files probably should not be compressed. I checked and if we get a random string of 94 different ASCII characters, I need at least ~450-500 bytes before I get a reduction to 90% of the original size. 

![90percent gain](https://github.com/aiidateam/disk-objectstore/assets/16132028/d1530d42-d081-4835-be71-39d6bb9a32fb)

Note however that a "real" file compresses better than a random sequence of characters, and also that for every CalcJob there are at least three small files (`_aiidasubmit.sh` and the two files in `.aiida`) that are stored and are just a few hundred bytes large, but still amount to a large fraction of the storage since there are many, and they can be easily compressed to 50%-70% of their original size since they are small.
So I need to decide a threshold

Reply:
From the analysis below, I'm tempted to let it compress unless it's zero bytes - we can improve in the future